version: '3.8'

services:
  database:
    image: postgres:16.0
    environment:
      POSTGRES_DB: 'main'
      POSTGRES_USER: 'admin'
      POSTGRES_PASSWORD: 'test'
      PGDATA: '/var/lib/postgresql/data/pgdata'
    restart: always
    expose:
      - "5432:5432"
    volumes:
      - ./volumes/database/data:/var/lib/postgresql/data/pgdata
      - ./volumes/database/postgresql.conf:/etc/postgresql.conf
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U admin -d main"]
      interval: 10s
      timeout: 5s
      retries: 5
      start_period: 10s

  database_admin:
    image: dpage/pgadmin4:7.2
    environment:
      PGADMIN_DEFAULT_EMAIL: "admin@pgadmin.com"
      PGADMIN_DEFAULT_PASSWORD: "test"
    volumes:
      - ./volumes/pgadmin/data:/var/lib/pgadmin
    ports:
      - "16111:80"
    restart: unless-stopped
    depends_on:
      database:
        condition: service_healthy

  nginx:
    image: nginx:latest
    ports:
      - '16112:16112'
    restart: always
    volumes:
      - ./volumes/nginx/nginx.conf:/etc/nginx/conf.d/default.conf
      - ./volumes/nginx/nginx.conf:/etc/nginx/nginx.conf:ro
      - ./volumes/nginx/sites-enabled:/etc/nginx/sites-enabled:ro
      - ./volumes/nginx/ssl:/etc/nginx/ssl:ro
      - ./volumes/nginx/var/www/:/var/www/
    depends_on:
      auth:
        condition: service_started

  auth:
    build: 
      context: ./
      dockerfile: ./auth/Dockerfile
    restart: on-failure
    command: uvicorn main:my_app --reload --host 0.0.0.0 --port 16112 --log-config ./log.ini
    environment:
      DB_HOST: 'database'
      DB_PORT: 5432
      DB_USER: 'admin'
      DB_PASSWORD: 'test'
      DB_MAIN: 'main'
      DB_TEST: 'test'
      TESTING: false
      PREFIX: '/api/v1'
    volumes:
      - ./volumes/auth/app.log:/app/app.log
    expose:
      - '16112:16112'
    depends_on:
      database:
        condition: service_healthy 